src/core/parse_worker.hpp

```diff
@@ -0,0 +1,132 @@
+#pragma once
+
+#include <string>
+#include <unordered_set>
+#include <vector>
+#include <nlohmann/json.hpp>
+#include "symbol.hpp"
+#include "token.hpp"
+
+namespace docgen {
+namespace core {
+
+class ParseWorker;
+
+using symbol_t = Symbol;
+using token_t = Token<symbol_t>;
+using routine_t = void (*)(const token_t& t);
+using worker_t = ParseWorker;
+
+/*
+ * A ParseWorker object holds an ordered array of SymbolHandler objects;
+ * at any given point, only one of these SymbolHandlers is being handled.
+ *
+ * On every call to proc(), the current handler is checked for a match;
+ * so long as it isn't matched (waiting to match), proc() will call for
+ * the handler to handle_workers(), which will process based on
+ * its array of ParseWorker objects. When the current handler is matched,
+ * proc() will call for the handler to handle_token(), which executes
+ * the routine set by handler's passed function pointer; then, set to
+ * move on to the next handler. On the last handler, this will wrap around
+ * to the first, for so long as the passed "iters" limiter isn't
+ * exceeded (unlimited by default); once it is, this ParseWorker is said
+ * to be in a "done" state, on which proc() will no longer execute.
+ *
+ * Said to be in a "working" state when handling a SymbolHandler after
+ * the first; this only matters to a SymbolHandler which is holding
+ * this ParseWorker.
+ */
+class ParseWorker 
+{
+	public:
+		struct SymbolHandler;
+
+		using handler_t = SymbolHandler;
+		using handlers_t = std::vector<handler_t>;
+		using handlers_init_t = std::initializer_list<handler_t>;
+
+		using workers_t = std::vector<worker_t>;
+		using workers_init_t = std::initializer_list<worker_t>;
+
+		/*
+		 * A SymbolHandler object holds an unordered set of symbols to match against (symbols_),
+		 * a function pointer which references a routine to execute on match (on_symbol_), and
+		 * an array of ParseWorker objects to process while still unmatched (workers_).
+		 *
+		 * If neg_ is set true, a given symbol will only match if it is not in symbols_.
+		 *
+		 * If working_ is not a nullptr, it indicates that one of the workers is currently in
+		 * the "working" state, and points to this worker; until it is done, only this worker
+		 * will be processed.
+		 */
+		class SymbolHandler
```

Comment:
This can be a separate class and not nested. It's easier to maintain if this is possible. Generally, a nested class is needed because it handles some internal data of outside class that can be further encapsulated. An instance of `SymbolHandler` does not manage any resources of an existing `ParseWorker` object. Conceptually, the existence of a `SymbolHandler` object is not conditioned on the existence of a `ParseWorker` object.

Reply:
I don't think this is necessarily true—to work with the outside class's members, it would have to be declared a `friend`, so I think there must be some context where omitting this is acceptable. However, I could kinda go either way on this, so I'll just put it outside the class, but still in `parse_worker.hpp` I think.

Reply:
A symbol handler can expose public api to modify its internal data so that a parseworker can just call them. I usually like to avoid nested classes and friends whenever possible because the simpler separated-class design almost always works.